File: greencompile-Intel/benchmarks/string_and_text_processing/sha1_pure_implementation/go/main.go
package main

import "encoding/binary"

const rounds = 20000

func rol(v uint32, n uint) uint32 {
	return (v << n) | (v >> (32 - n))
}

func sha1(msg []byte) [20]byte {
	h0 := uint32(0x67452301)
	h1 := uint32(0xEFCDAB89)
	h2 := uint32(0x98BADCFE)
	h3 := uint32(0x10325476)
	h4 := uint32(0xC3D2E1F0)

	bitLen := uint64(len(msg)) * 8
	total := len(msg) + 1 + 8
	pad := (64 - (total % 64)) % 64
	total += pad

	block := make([]byte, 64)
	for off := 0; off < total; off += 64 {
		for i := 0; i < 64; i++ {
			idx := off + i
			if idx < len(msg) {
				block[i] = msg[idx]
			} else if idx == len(msg) {
				block[i] = 0x80
			} else if idx < total-8 {
				block[i] = 0x00
			} else {
				shift := uint((total - 1 - idx) * 8)
				block[i] = byte(bitLen >> shift)
			}
		}

		var w [80]uint32
		for i := 0; i < 16; i++ {
			w[i] = binary.BigEndian.Uint32(block[i*4 : i*4+4])
		}
		for i := 16; i < 80; i++ {
			w[i] = rol(w[i-3]^w[i-8]^w[i-14]^w[i-16], 1)
		}

		a := h0
		b := h1
		c := h2
		d := h3
		e := h4

		for i := 0; i < 80; i++ {
			var f, k uint32
			switch {
			case i < 20:
				f = (b & c) | (^b & d)
				k = 0x5A827999
			case i < 40:
				f = b ^ c ^ d
				k = 0x6ED9EBA1
			case i < 60:
				f = (b & c) | (b & d) | (c & d)
				k = 0x8F1BBCDC
			default:
				f = b ^ c ^ d
				k = 0xCA62C1D6
			}
			temp := rol(a, 5) + f + e + k + w[i]
			e = d
			d = c
			c = rol(b, 30)
			b = a
			a = temp
		}

		h0 += a
		h1 += b
		h2 += c
		h3 += d
		h4 += e
	}

	var out [20]byte
	binary.BigEndian.PutUint32(out[0:4], h0)
	binary.BigEndian.PutUint32(out[4:8], h1)
	binary.BigEndian.PutUint32(out[8:12], h2)
	binary.BigEndian.PutUint32(out[12:16], h3)
	binary.BigEndian.PutUint32(out[16:20], h4)
	return out
}

func main() {
	msg := []byte("GreenCompile SHA-1 pure implementation benchmark input.")
	var checksum byte
	for i := 0; i < rounds; i++ {
		out := sha1(msg)
		checksum ^= out[i%20]
	}
	_ = checksum
}

File: greencompile-Intel/benchmarks/string_and_text_processing/sha1_pure_implementation/rust/main.rs
const ROUNDS: usize = 20000;

fn rol(v: u32, n: u32) -> u32 {
    (v << n) | (v >> (32 - n))
}

fn sha1(msg: &[u8]) -> [u8; 20] {
    let mut h0: u32 = 0x67452301;
    let mut h1: u32 = 0xEFCDAB89;
    let mut h2: u32 = 0x98BADCFE;
    let mut h3: u32 = 0x10325476;
    let mut h4: u32 = 0xC3D2E1F0;

    let bit_len: u64 = (msg.len() as u64) * 8;
    let mut total = msg.len() + 1 + 8;
    let pad = (64 - (total % 64)) % 64;
    total += pad;

    let mut block = [0u8; 64];
    let mut off = 0usize;
    while off < total {
        for i in 0..64 {
            let idx = off + i;
            if idx < msg.len() {
                block[i] = msg[idx];
            } else if idx == msg.len() {
                block[i] = 0x80;
            } else if idx < total - 8 {
                block[i] = 0x00;
            } else {
                let shift = (total - 1 - idx) * 8;
                block[i] = (bit_len >> shift) as u8;
            }
        }

        let mut w = [0u32; 80];
        for i in 0..16 {
            let j = i * 4;
            w[i] = ((block[j] as u32) << 24)
                | ((block[j + 1] as u32) << 16)
                | ((block[j + 2] as u32) << 8)
                | (block[j + 3] as u32);
        }
        for i in 16..80 {
            w[i] = rol(w[i - 3] ^ w[i - 8] ^ w[i - 14] ^ w[i - 16], 1);
        }

        let mut a = h0;
        let mut b = h1;
        let mut c = h2;
        let mut d = h3;
        let mut e = h4;

        for i in 0..80 {
            let (f, k) = if i < 20 {
                ((b & c) | (!b & d), 0x5A827999)
            } else if i < 40 {
                (b ^ c ^ d, 0x6ED9EBA1)
            } else if i < 60 {
                ((b & c) | (b & d) | (c & d), 0x8F1BBCDC)
            } else {
                (b ^ c ^ d, 0xCA62C1D6)
            };
            let temp = rol(a, 5)
                .wrapping_add(f)
                .wrapping_add(e)
                .wrapping_add(k)
                .wrapping_add(w[i]);
            e = d;
            d = c;
            c = rol(b, 30);
            b = a;
            a = temp;
        }

        h0 = h0.wrapping_add(a);
        h1 = h1.wrapping_add(b);
        h2 = h2.wrapping_add(c);
        h3 = h3.wrapping_add(d);
        h4 = h4.wrapping_add(e);

        off += 64;
    }

    let mut out = [0u8; 20];
    let h = [h0, h1, h2, h3, h4];
    for i in 0..5 {
        out[i * 4] = (h[i] >> 24) as u8;
        out[i * 4 + 1] = (h[i] >> 16) as u8;
        out[i * 4 + 2] = (h[i] >> 8) as u8;
        out[i * 4 + 3] = h[i] as u8;
    }
    out
}

fn main() {
    let msg = b"GreenCompile SHA-1 pure implementation benchmark input.";
    let mut checksum: u8 = 0;
    for i in 0..ROUNDS {
        let out = sha1(msg);
        checksum ^= out[i % 20];
    }
    if checksum == 0 {
        print!("");
    }
}
